modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/playback/SleepTimer.kt

```diff
@@ -31,171 +30,191 @@ class SleepTimer @Inject constructor(
         const val TAG: String = "SleepTimer"
     }
 
-    private var sleepTimeMs: Long? = null
-    private var lastSleepAfterTime: Duration? = null
-    private var lastSleepAfterEndOfChapterTime: Duration? = null
-    private var lastTimeSleepTimeHasFinished: Duration? = null
-    private var lastEpisodeUuidAutomaticEnded: String? = null
+    private var sleepTimerHistory = SleepTimerHistory()
 
-    fun sleepAfter(duration: Duration, onSuccess: () -> Unit) {
-        val sleepAt = System.currentTimeMillis().milliseconds + duration
+    private val _stateFlow: MutableStateFlow<SleepTimerState> = MutableStateFlow(SleepTimerState())
+    val stateFlow: StateFlow<SleepTimerState> = _stateFlow
+
+    fun getState(): SleepTimerState = _stateFlow.value
 
-        if (createAlarm(sleepAt.inWholeMilliseconds)) {
-            lastSleepAfterTime = duration
-            cancelAutomaticSleepOnEpisodeEndRestart()
-            cancelAutomaticSleepOnChapterEndRestart()
-            onSuccess()
+    fun updateSleepTimerStatus(
+        sleepTimeRunning: Boolean,
+        sleepAfterEpisodes: Int = 0,
+        sleepAfterChapters: Int = 0,
+        timeLeft: Duration? = null,
+    ) {
+        _stateFlow.update { currentState ->
+            currentState.copy(
+                isSleepTimerRunning = sleepTimeRunning,
+                numberOfEpisodesLeft = sleepAfterEpisodes,
+                numberOfChaptersLeft = sleepAfterChapters,
+                timeLeft = if (!sleepTimeRunning || sleepAfterEpisodes != 0 || sleepAfterChapters != 0) {
+                    ZERO
+                } else {
+                    timeLeft ?: currentState.timeLeft
+                },
+            )
         }
     }
 
-    fun addExtraTime(minutes: Int) {
-        val currentTimeMs = sleepTimeMs
-        if (currentTimeMs == null || currentTimeMs < 0) {
+    fun sleepAfter(duration: Duration, onSuccess: () -> Unit) {
+        updateSleepTimerStatus(sleepTimeRunning = true, timeLeft = duration)
+
+        cancelAutomaticSleepOnEpisodeEndRestart()
+        cancelAutomaticSleepOnChapterEndRestart()
+
+        sleepTimerHistory = sleepTimerHistory.copy(
+            lastSleepAfterTime = duration,
+            lastTimeSleepTimeHasFinished = System.currentTimeMillis().milliseconds + duration,
+        )
```

Comment:
Do you think this could be cleaner if `SleepTimerHistory` was a sealed interface? Something like the following:

```suggestion
        sleepTimerHistory = SleepTimerHistory.AfterTime(
            lastSleepAfterTime = duration,
            lastTimeSleepTimeHasFinished = System.currentTimeMillis().milliseconds + duration,
        )
```

```
sealed interface SleepTimerHistory {
    data class AfterTime(
        val lastSleepAfterTime: Duration,
        val lastTimeSleepTimeHasFinished: Duration,
    ) : SleepTimerHistory
    data object AfterChapter : SleepTimerHistory
    data object AfterEpisode : SleepTimerHistory
}
```

Was there a reason for using the `copy` method rather than creating the class again?

Reply:
I updated to sealed interface. It looks better now